### PR TITLE
Fix heartbeats being counted as a logical WFT

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
WFT heartbeats should not count as a "real" WFT, they should instead be ignored when scanning ahead to the next logical WFT.

## Why?
Treating heartbeats as a full logical WFT meant that when scanning ahead to pre-resolve local activities, they were not always resolved when they should have been. This led to weird nondeterministic behavior where we would try to newly execute LAs that should have already been resolved.

## Checklist

1. Closes 

2. How was this tested:
Added repro test cases, and one long regression integ test that is not on by default.

Will also manually run tests in lang sdks w/ updated core.

3. Any docs updates needed?

### CODE REVIEW

File: core/src/worker/workflow/history_update.rs [hunk 1]
```diff
@@ -366,6 +400,46 @@ pub mod tests {
         assert_eq!(seq_2.last().unwrap().event_id, 8);
     }
 
+    #[tokio::test]
+    async fn history_ends_abruptly() {
```
Comment (bergundy) [inline core/src/worker/workflow/history_update.rs:404]: Do we have a test for WFT failed and timeout?
I was too lazy too look.

File: tests/integ_tests/workflow_tests/local_activities.rs [hunk 2]
```diff
@@ -632,3 +637,139 @@ async fn repro_nondeterminism_with_timer_bug() {
...
+        activity_type: "delay".to_string(),
+        start_to_close_timeout: Some(Duration::from_secs(20)),
+        input: "hi!".as_json_payload().expect("serializes fine"),
+        ..Default::default()
+    })
+    .await;
+    Ok(().into())
+}
+
+// Expensive to run - worth enabling on a stress/regression pipeline.
```
Comment (bergundy) [inline tests/integ_tests/workflow_tests/local_activities.rs:666]: Can we make this cheaper by reducing the number iterations and cache size?

**Review by bergundy — APPROVED** [review]

File: core/src/worker/workflow/history_update.rs (hunk 1 above)
Comment (Sushisource) [reply core/src/worker/workflow/history_update.rs:404]: Yes

File: tests/integ_tests/workflow_tests/local_activities.rs (hunk 2 above)
Comment (Sushisource) [reply tests/integ_tests/workflow_tests/local_activities.rs:666]: Maybe, but we have a need for a nightly-esque core pipeline anyway. I'd prefer to just make that.

**Comment by Sushisource** [comment]
After continuing to hammer away with weird tests I managed to trigger a panic around expecting run state to not suddenly vanish. Don't think it has anything to do with these changes, but obviously I want to fix that.

**Review by Spikhalskiy — APPROVED** [review]

**Pull request merged**